File: src/Tree/BinaryTreeBFSGuide/BinaryTreeZigzagLevelOrderTraversal/BinaryTreeZigzagLevelOrderTraversal.java
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        
        if(root == null) return res;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        
        boolean FL = true;
        
        while(!q.isEmpty()){
            int size = q.size();
            List<Integer> sub = new LinkedList<>();
            
            for(int i = 0; i < size; i++){
                TreeNode first = q.poll();
                if(FL){
                    sub.add(first.val);
                }else{
                    sub.add(0, first.val);
                }
                if(first.left != null){
                    q.add(first.left);
                }
                
                if(first.right != null){
                    q.add(first.right);
                }
            }
            res.add(sub);
            FL = !FL;
        }
        return res;
    }
}

File: src/Tree/BinaryTreeBFSGuide/BinaryTreeZigzagLevelOrderTraversal/BinaryTreeZigzagLevelOrderTraversal.py
from collections import deque
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        
        if root is None:
            return res
        
        queue = deque([root])
        
        is_left_to_right = True
        
        while len(queue) != 0:
            size = len(queue)
            
            level_list = []
            
            for idx in range(size):
                cur_node = queue.popleft()
                
                if is_left_to_right:
                    level_list.append(cur_node.val)
                else:
                    level_list.insert(0, cur_node.val)
                
                if cur_node.left is not None:
                    queue.append(cur_node.left)
                
                if cur_node.right is not None:
                    queue.append(cur_node.right)
            
            is_left_to_right = not is_left_to_right
            
            res.append(level_list)
        
        return res
